README.md
node-chatrooms
==============

Like the example Node.js chat application except it supports multiple rooms and
uses Socket.IO instead of long-polling. This is just a technical excercise.

TODO
----

* Name rooms
* Option to add a password to rooms when creating
* Actual authentication/logins
* Check compatibility with different browsers
* Add comments to code
* Add tests
* Room roles (Owner, moderator, user)
* Organize client code with Backbone.js
* Fix the error messages when picking a nickname
* Handle sessions better
* Fix socket.io problems with Chrome
* Update socket.io to development version

HISTORY.md
0.0.8, 2011/05/18
-----------------

* Fixed the problem where messages would be sent to all users
* Saves chat rooms and messages in Mongo with Mongoose

0.0.6, 2011/05/18
-----------------

* Refactored code to make things cleaner
* Created seperate files for urls, views, and models (kinda like Django)

0.0.5, 2011/05/17
-----------------

* Changed code from long-polling to leverage Socket.IO
* Refactored a lot of code

0.0.2, 2011/05/17
-----------------

* Changed some methods that altered data to POST instead of GET
* Refactored some code to get parameters using Express helpers

0.0.1, 2011/05/16
-----------------

* Initial commit
* Working using long-polling, node.js, and Express

app.js
var express = require('express'),
	mongoose = require('mongoose'),
	path = require('path');

var app = express.createServer();

app.configure(function() {
	app.set('views', path.join(__dirname, 'templates'));
	app.set('view engine', 'jade');
	app.use(express.bodyParser());
	app.use(express.methodOverride());
	app.use(require('stylus').middleware({ src: path.join(__dirname, 'public') }));
	app.use(app.router);
	app.use(express.static(path.join(__dirname, 'public')));
});

app.configure('development', function(){
	app.use(express.errorHandler({ dumpExceptions: true, showStack: true })); 
});

app.configure('production', function(){
	app.use(express.errorHandler()); 
});

this.app = app;
this.socket = require('socket.io').listen(app);

var urls = require('./urls');

for (var name in urls) {
	var url = urls[name];
	if (!url.method) {
		url.method = 'get';
	}
	if (!(url.view instanceof Array)) {
		var args = [url.path, url.view];
	} else {
		url.view.unshift(url.path);
		var args = url.view;
	}
	app[url.method].apply(app, args);
}
mongoose.connect('mongodb://localhost/test');

if (!module.parent) {
	app.listen(3000);
	console.log("Express server listening on port %d", app.address().port);
}

rooms.js
var utils = require('./utils');

var MESSAGE_BACKLOG = 1000,
	SESSION_TIMEOUT = 1 * 60 * 1000, // 2 minute
	CALLBACK_TIMEOUT = 2 * 60 * 1000; // 2 minute
	
function validateNickname(nickname) {
	if (nickname.length <= 0) {
		return 'Nickname must not be blank.';
	} else if (nickname.length > 50) {
		return 'Nickname can not be longer than 50 characters.';
	} else if (/[^\w-]/.test(nickname)) {
		return 'Bad character found in nickname. Only letters, numbers, "-", and "_" are allowed';
	} else {
		return null;
	}
}

function generateId(length) {
	var id = '';
	var chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	
	length = length || 26;
	
	for (var i=0; i < length; i++) {
		id += chars.charAt(Math.floor(Math.random() * chars.length));
	}
	return id;
}

function Room(id) {
	this.id = id;
	this.messages = [];
	this.sessions = {};
	this.callbacks = [];
	this.created = new Date();
}

Room.prototype.leave = function(sessionId) {
	var session = this.sessions[sessionId];
	if (!session) {
		return { error: 'No session exists with that ID.', status: 400 }
	}
	this.addMessage(session.nickname, 'left', 'left');
	delete this.sessions[sessionId];
	return null;
};

Room.prototype.join = function(nickname) {
	nickname = utils.trim(nickname);
	
	if (nickname in this.sessions) {
		return { error: 'Nickname already exists. Choose another one.', status: 400 };
	}
	
	var error = validateNickname(nickname);
	if (error) {
		return { error: error, status: 400 };
	}
	
	var id = generateId();
	while (id in this.sessions) {
		id = generateId();
	}
	var session = this.sessions[id] = {
		id: id,
		nickname: nickname,
		lastActive: new Date(),
		poke: function() {
			session.lastActive = new Date();
		}
	}
	
	this.addMessage(nickname, 'joined', 'joined');
	
	return session;
};

Room.prototype.addMessage = function(nickname, type, text) {
	
	switch (type) {
		case 'msg':
			console.log('<' + nickname + '> ' + text);
			break;
		case 'joined':
			console.log(nickname + ' joined');
			break;
		case 'left':
			console.log(nickname + ' left');
			break;
	}
	
	var message = {
		nickname: nickname,
		type: type,
		text: text,
		timestamp: (new Date()).getTime()
	};
	
	this.messages.push(message);
	
	while (this.callbacks.length > 0) {
		this.callbacks.shift().callback([message]);
	}
	
	while (this.messages.length > MESSAGE_BACKLOG) {
		this.messages.shift();
	}
};

Room.prototype.getMessages = function(since, callback) {
	var matching = [];
	for (var i=0; i < this.messages.length; i++) {
		var message = this.messages[i];
		if (message.timestamp > since) {
			matching.push(message);
		}
	}
	
	if (matching.length != 0) {
		callback(matching);
	} else {
		this.callbacks.push({
			timestamp: new Date(),
			callback: callback
		});
	}
};

Room.prototype.who = function() {
	var nicknames = [];
	for (var id in this.sessions) {
		nicknames.push(this.sessions[id].nickname);
	}
	return nicknames;
}

var rooms = {};

function createRoom() {
	
	// Make sure that the room ID is unique
	var id = generateId();
	while(id in rooms) {
		id = generateId();
	}
	
	rooms[id] = new Room(id);
	return id;
}


setInterval(function() {
	var now = new Date();
	for (var roomId in rooms) {
		var room = rooms[roomId];
		while (room.callbacks.length > 0 && now - room.callbacks[0].timestamp > CALLBACK_TIMEOUT) {
			room.callbacks.shift().callback([]);
		}
	}
}, 3000);

setInterval(function() {
	var now = new Date();
	for (var roomId in rooms) {
		var room = rooms[roomId];
		for (var sessionId in room.sessions) {
			var session = room.sessions[sessionId];
			if (now - session.lastActive > SESSION_TIMEOUT) {
				room.leave(sessionId);
			}
		}
	}
}, 1000);

// Views

this.views = {}

this.views.index = function(req, res) {
	console.log('Showing ' + Object.keys(rooms).length + ' rooms');
	res.render('index', {
		rooms: rooms,
		title: 'node chat'
	});
};

this.views.create = function(req, res) {
	var id = createRoom();
	res.redirect('/' + id);
};

this.views.show = function(req, res) {
	console.log('Showing room ' + req.room.id);
	res.render('show', {
		room: req.room,
		title: 'node chat'
	});
};

this.views.join = function(req, res) {
	res.contentType('json');
	var session = req.room.join(req.param('nickname'));
	if (session.error) {
		res.send(session.error, session.status);
		return;
	}
	
	res.send(session, 200);
};

this.views.leave = function(req, res) {
	var sessionId = req.param('sessionId');
	var response = req.room.leave(sessionId);
	if (response) {
		res.contentType('json');
		res.send(response.error, response.status);
	} else {
		res.redirect('/');
	}
	
};

this.views.send = function(req, res) {
	res.contentType('json');
	var sessionId = req.param('sessionId');
	var text = req.param('text');
	var session = req.room.sessions[sessionId];
	session.poke();
	if (!session || !text) {
		res.send('No such session id', 400);
		return;
	}
	req.room.addMessage(session.nickname, 'msg', text);
	res.send({ success: 'success' }, 200);
};

this.views.receive = function(req, res) {
	res.contentType('json');
	var since = req.param('since');
	if (!since) {
		res.send('Must supply since parameter', 400);
		return;
	}
	var sessionId = req.param('sessionId');
	var session;
	if (sessionId && req.room.sessions[sessionId]) {
		session = req.room.sessions[sessionId];
		session.poke();
	}
	
	since = parseInt(since, 10);
	
	req.room.getMessages(since, function(messages) {
		if (session) session.poke();
		res.send({ messages: messages }, 200);
	});
};

this.views.who = function(req, res) {
	res.contentType('json');
	res.send({
		nicknames: req.room.who(),
	}, 200);
};

this.middleware = {}

this.middleware.getRoomOr404 = function(req, res, next) {
	var room = rooms[req.params.id];
	if (room) {
		req.room = room;
		next();
	} else {
		next(new Error('Room not found with an ID of ' + req.params.id));
	}
};



utils.js
this.trim = function(stringToTrim) {
	return stringToTrim.replace(/^\s+|\s+$/g, '');
};

this.ltrim = function(stringToTrim) {
	return stringToTrim.replace(/^\s+/, '');
};

this.rtrim = function(stringToTrim) {
	return stringToTrim.replace(/\s+$/, '');
};

this.urlRE = /https?:\/\/([-\w\.]+)+(:\d+)?(\/([^\s]*(\?\S+)?)?)?/g,

this.escape = function(inputHtml) {
	inputHtml = inputHtml.toString();
	return inputHtml.replace(/&/g, '&amp;')
		.replace(/</g, '&lt;')
		.replace(/>/g, '&gt;');
};

this.zeroPad = function(digits, n) {
	n = n.toString();
	while (n.length < digits) {
		n = '0' + n;
	}
	return n;
};

this.timeString = function(date) {
	var minutes = date.getMinutes().toString();
	var hours = date.getHours().toString();
	return this.zeroPad(2, hours) + ':' + this.zeroPad(2, minutes);
};

this.isBlank = function(text) {
	var blank = /^\s*$/;
	return (text.match(blank) !== null);
};

public/js/client.js
var CONFIG = {
	debug: false,
	roomId: $('#messages').attr('rev'),
	nickname: '#',
	sessionId: null,
	lastMessageTime: (new Date()).getTime(),
	unread: 0
};

var $messageTemplate = $('#message-template').clone().removeAttr('id').remove(),
	title = document.title,
	nicknames = [],
	startTime;
	
function dateToRelativeTime(date, nowThreshold) {
	var delta = new Date() - date;

	nowThreshold = parseInt(nowThreshold, 10);

	if (isNaN(nowThreshold)) {
		nowThreshold = 0;
	}

	if (delta <= nowThreshold) {
		return 'Just now';
	}

	var units = null;
	var conversions = {
		millisecond: 1, // ms -> ms
		second: 1000, // ms -> sec
		minute: 60, // sec -> min
		hour: 60, // min -> hour
		day: 24, // hour -> day
		month: 30, // day -> month (roughly)
		year: 12 // month -> year
	};

	for (var key in conversions) {
		if (delta < conversions[key]) {
			break;
		} else {
			units = key; // keeps track of the selected key over the iteration
			delta = delta / conversions[key];
		}
	}

	// pluralize a unit when the difference is greater than 1.
	delta = Math.floor(delta);
	if (delta !== 1) { units += "s"; }
	return [delta, units].join(" ");
};

function dateFromString(str) {
	return new Date(Date.parse(str));
};

util = {
	urlRE: /https?:\/\/([-\w\.]+)+(:\d+)?(\/([^\s]*(\?\S+)?)?)?/g,
	escape: function(inputHtml) {
		inputHtml = inputHtml.toString();
		return inputHtml.replace(/&/g, '&amp;')
			.replace(/</g, '&lt;')
			.replace(/>/g, '&gt;');
	},
	zeroPad: function(digits, n) {
		n = n.toString();
		while (n.length < digits) {
			n = '0' + n;
		}
		return n;
	},
	timeString: function(date) {
		var minutes = date.getMinutes().toString();
		var hours = date.getHours().toString();
		return this.zeroPad(2, hours) + ':' + this.zeroPad(2, minutes);
	},
	isBlank: function(text) {
		var blank = /^\s*$/;
		return (text.match(blank) !== null);
	}
};

function scrollDown() {
	window.scrollBy(0, 100000000000000000);
	$('#entry').focus();
}

function addMessage(message) {
	
	if (message.text === null) return;
	
	var $message = $messageTemplate.clone();
	
	text = util.escape(message.text);
	
	var nicknameRe = new RegExp(CONFIG.nickname);
	if (nicknameRe.exec(message.text)) {
		$message.addClass('personal');
	}
	
	message.text = message.text.replace(util.urlRE, '<a target="_blank" href="$&">$&</a>');
	
	if (message.type == 'msg' && !hasFocus()) {
		CONFIG.unread++;
	} else if (message.type == 'joined' || message.type == 'left') {
		$('span', '#users').text(message.who.length);
	}
	console.log(message);
	$message.addClass(message.type)
		.find('.timestamp').html(util.timeString(new Date())).end()
		.find('.nickname').html(util.escape(message.nickname)).end()
		.find('.text').html(message.text);
	
	$('#messages').append($message);
	scrollDown();
}

function clearUnread() {
	CONFIG.unread = 0;
	updateTitle();
}

function updateTitle() {
	if (CONFIG.unread) {
		document.title = '(' + CONFIG.unread.toString() + ') ' + title;
	} else {
		document.title = title;
	}
}

function updateUptime() {
	if (startTime) {
		$('#uptime').text(dateToRelativeTime(startTime));
	}
}

function onConnect(session) {
	CONFIG.nickname = session.nickname
	
	showChat();
	
	var socket = new io.Socket(null, {port: 3000});
	socket.connect();
	socket.on('message', function(message) {
		switch (message.type) {
			case 'init':
				CONFIG.sessionId = message.sessionId;
				socket.send({
					type: 'joined',
					nickname: CONFIG.nickname,
					roomId: CONFIG.roomId,
					sessionId: CONFIG.sessionId
				});
				break;
			case 'msg':
			case 'joined':
			case 'left':
				addMessage(message);
				break;
		}
	});

	$('#talk').submit(function(e) {
		e.preventDefault();
		var $entry = $('#entry');
		var message = $entry.val();
		$entry.val('');
		
		if (message.length > 0) {
			socket.send({
				type: 'msg',
				text: message,
				sessionId: CONFIG.sessionId,
				roomId: CONFIG.roomId,
				nickname: CONFIG.nickname
			});
		}
	});
	
	var hasLeft = false;
	$(window).bind('beforeunload unload', function() {
		if (!hasLeft) {
			socket.send({
				type: 'left',
				nickname: CONFIG.nickname,
				roomId: CONFIG.roomId,
				sessionId: CONFIG.sessionId
			});
		}
	});
}

function join(nickname) {
	
	$.ajax({
		type: 'post',
		cached: false,
		url: '/' + CONFIG.roomId + '/join',
		data: {
			nickname: nickname
		},
		dataType: 'json',
		error: function(data) {
			alert(data.responseText);
			showConnect();
		},
		success: onConnect
	});
	
}

function hasFocus() {
	return $('#entry').is(':focus');
}

function showConnect() {

	$('#connect').show();
	$('#loading').hide();
	$('#chat').hide();
	$('#nickname').focus();

}

function showChat() {

	$('#connect').hide();
	$('#loading').hide();
	$('#chat').show();
	$('#entry').focus();
	scrollDown();
	
}

function showLoading() {

	$('#connect').hide();
	$('#chat').hide();
	$('#loading').show();
	
}

$(function() {

	$('span.timestamp', '#messages').each(function() {
		var time = parseInt($(this).text());
		var dateString = util.timeString(new Date(time));
		$(this).text(dateString);
	});
	
	startTime = new Date($('#uptime').text());
	
	updateUptime();
	showConnect();
	
	setInterval(function() {
		updateUptime();
	}, 25 * 1000);
	
	$('#entry').mousedown(clearUnread);
	$('#join').submit(function(e) {
		e.preventDefault();
		showLoading();
		var nickname = $('#nickname').val();
		join(nickname);
	});
	
	$('#users').click(function(e) {
		e.preventDefault();
		who();
		var nicknamesString = '(none)';
		if (nicknames.length > 0) {
			nicknamesString = nicknames.join(', ');
		}
		addMessage('notice', new Date(), 'users:', nicknamesString);
	});
	
});


views.js
var socket = require('./app').socket,
	Room = require('./models').Room;
	
function generateId(length) {
	var id = '';
	var chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

	length = length || 26;

	for (var i=0; i < length; i++) {
		id += chars.charAt(Math.floor(Math.random() * chars.length));
	}
	return id;
}

this.getRoomOr404 = function(req, res, next) {
	Room.findById(req.param('id'), function(error, room) {
		if (room) {
			req.room = room;
			next();
		} else {
			next(new Error('Room not found with an ID of ' + req.params.id));
		}
	});
};

var roomSessions = {};

var nicknames = function(roomId) {
	var nicknames = [];
	for (var id in roomSessions[roomId]) {
		nicknames.push(roomSessions[roomId][id].nickname);
	}
	return nicknames;
}
	
var who = function(roomId) {
	var sessionIds = [];
	for (var id in roomSessions[roomId]) {
		sessionIds.push(id);
	}
	return sessionIds;
}

Room.find({}, function(err, rooms) {
	for (var i=0; i < rooms.length; i++) {
		var room = rooms[i];
		roomSessions[room.id] = {}
	}
});

this.rooms = {
	index: function(req, res) {
		Room.find({}, function(err, rooms) {
			console.log('Showing ' + Object.keys(rooms).length + ' rooms');
			res.render('index', {
				rooms: rooms,
				title: 'node chat'
			});
		});
	},
	create: function(req, res) {
		var room = new Room();
		room.save();
		roomSessions[room.id] = {}
		res.redirect('/' + room.id);
	},
	show: function(req, res) {
		console.log('Showing room ' + req.room.id);
		res.render('show', {
			room: req.room,
			title: 'node chat'
		});
	},
	join: function(req, res) {
		var error = null;
		var nickname = req.param('nickname');
		if (nickname.length <= 0) {
			error = 'Nickname must not be blank.';
		} else if (nickname.length > 50) {
			error = 'Nickname can not be longer than 50 characters.';
		} else if (/[^\w-]/.test(nickname)) {
			error = 'Bad character found in nickname. Only letters, numbers, "-", and "_" are allowed';
		} else {
			for (var sessionId in roomSessions[req.room.id]) {
				var session = roomSessions[req.room.id];
				if (session.nickname === nickname) {
					error = 'Nickname already exists. Pick another one.';
				}
			}
		}
		if (error) {
			res.send({ error: error }, 400);
		} else {
			res.send({ nickname: nickname }, 200);
		}
	}
};

socket.on('connection', function(client) {

	client.send({ type: 'init', sessionId: client.sessionId });
	
	client.on('message', function(message) {
		Room.findById(message.roomId, function(error, room) {
			var session = roomSessions[room.id][message.sessionId];
			var msg = { type: message.type }
			if (message.type === 'joined' && !session) {
				if (!session) {
					session = roomSessions[room.id][message.sessionId] = {
						nickname: message.nickname,
						client: client
					}
				}
				msg.who = nicknames(room.id);
				msg.text = 'joined';
				msg.nickname = session.nickname;
			} else if (message.type === 'left') {
				msg.nickname = session.nickname;
				msg.who = nicknames(room.id);
				msg.text = 'left';
				delete roomSessions[room.id][message.sessionId];
			} else {
				msg.nickname = session.nickname;
				msg.text = message.text;
			}
			room.messages.push({
				type: msg.type,
				text: msg.text,
				nickname: msg.nickname,
				timestamp: Date.now()
			});
			room.save();
			for (var sessionId in roomSessions[room.id]) {
				roomSessions[room.id][sessionId].client.send(msg);
			}
		});
		//socket.broadcast(msg);
	});
	
});